### ISSUE TITLE
Segmentation fault (core dumped) when run tor in vanilla and priority

### ISSUE THREAD

**Issue by MHTTHM** [shadow/shadow-plugin-tor#50]
I have built shadow successfully,  but it failed when I run tor in vanilla and priority. 
```
shadow -w 3 shadow.config.xml > shadow.log
** Starting Shadow v1.12.0-104-gcb5afeb 2018-03-25 (built 2018-04-10) with GLib v2.48.2 and IGraph v0.7.1
Segmentation fault (core dumped)

```
I checked the core file, 
```
Core was generated by `shadow -w 3 shadow.config.xml --preload=/home/mason/.shadow/lib/libshadow-inter'.
Program terminated with signal SIGSEGV, Segmentation fault.
```
The system I use is Ubuntu server 16.04.

**Comment by jtracey** [comment]
Have you run shadow's tests? Also, what does the log file say?

**Comment by MHTTHM** [comment]
yes ! I have run the minimal example and it completed 40 get requests. 
The log file says:
```
00:00:00.112338 [thread-0] 00:00:01.000000000 [message] [server33~178.236.6.251] [shd-process.c:1050] [_process_start] starting process 'server33.tgen.1000'
00:00:00.126470 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126484 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126509 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126526 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126530 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126848 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126857 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126861 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126864 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126868 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126871 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126874 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126877 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126881 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126885 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126888 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126891 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126895 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126898 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126901 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126904 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126908 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126912 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126915 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126918 [thread-1] 00:00:01.000000000 [warning] [server12~183.79.135.206] [shd-process.c:4740] [process_emu___cxa_atexit] atexit at library close is not currently supported
00:00:00.126928 [thread-1] 00:00:01.000000000 [message] [server12~183.79.135.206] [shd-process.c:416] [_process_loadPlugin] process 'server12.tgen.1000' successfully loaded plugin 'tgen' at path '/home/mason/.shadow/lib/libshadow-plugin-tgen.so' into new namespace '0x7fa5cc08d430' in 0.023476 seconds
00:00:00.126978 [thread-1] 00:00:01.000000000 [message] [server12~183.79.135.206] [shd-process.c:484] [_process_loadPlugin] **found '(null)' at 0x7fa5c4200c31**
```
found 'null' at somewhere , it seems shadow have referenced to an empty address.there are 136 such errors in the log file.

**Comment by robgjansen** [comment]
Sorry, I'm confused about where the segfault is occurring if the minimal example succeeded. There are 3 steps:

  1. Did you run `./setup test` from the Shadow directory after installing? Did any of those test cases fail?
  2. Did you `cd resource/examples` and then `shadow -w 2 shadow.config.xml > shadow.log`. Did this test fail?
  3. Did you install the `shadow-plugin-tor` plugin (https://github.com/shadow/shadow-plugin-tor) and then run the `shadow-plugin-tor/resource/shadowtor-minimal` example? Did that fail?

Thanks.

**Comment by MHTTHM** [comment]
Thank you for your reply.
1. I have run `./setup test` from the shadow .It's 100 passed.
  ```
100% tests passed, 0 tests failed out of 66

Total Test time (real) =  13.86 sec
2018-04-17 11:13:11,710 [INFO] ctest -j4 returned 0
```
2. I have done this test successfully.
  ```
~/Tor/shadow/resource/examples$ shadow -w 2 shadow.config.xml > shadow.log
** Starting Shadow v1.12.0-104-gcb5afeb 2018-03-25 (built 2018-04-10) with GLib v2.48.2 and IGraph v0.7.1
** Stopping Shadow, returning code 0 (success)

```
3.yes, I have installed the `shadow-plugin-tor` plugin and run the minimal example, it works.
  ```
/Tor/shadow-plugin-tor/resource/shadowtor-minimal$ shadow -w 2 shadow.config.xml > shadow.log
** Starting Shadow v1.12.0-104-gcb5afeb 2018-03-25 (built 2018-04-10) with GLib v2.48.2 and IGraph v0.7.1
** Stopping Shadow, returning code 0 (success)

```
Then I did this:
```
cd resource
tar xaf shadowtor-toy-config.tar.xz
cp -R shadowtor-toy-config vanilla
cp -R shadowtor-toy-config priority
```
but when I did `shadow -w 3 shadow.config.xml > shadow.log` in vanilla , it failed.

Thanks.

**Comment by robgjansen** [comment]
Thank you for the update, that indeed helps us narrow down the problem. Could you post the tail (last 10 lines or so) of the `shadow.log` file?

The next step would be to run shadow like:
`shadow -w 3 -g shadow.config.xml > shadow.log`

Then Shadow will pause on startup and print out the process pid number. You can use gdb to attach to the process with `gdb --pid=$PID`. Once inside gdb, type `continue` (probably twice to start all threads). Once the segfault occurs, gdb will halt the process. Then you can type `backtrace` and post the results here.

Thank you for helping track this down!

**Comment by MHTTHM** [comment]
Thank you for your help! I think I will never solve the problem by myself. Here's my update.
the tail of the `shadow.log` file is like this:
```
00:00:20.152309 [thread-0] 00:02:31.000000000 [message] [webclient40~11.0.1.62] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,0,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
00:00:20.152319 [thread-0] 00:02:31.000000000 [message] [relaymiddle29~193.166.167.85] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,0,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
00:00:20.152329 [thread-0] 00:02:31.000000000 [message] [webclient44~11.0.1.63] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,0,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
00:00:20.152339 [thread-0] 00:02:31.000000000 [message] [webclient252~11.0.1.64] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,0,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
00:00:20.152346 [thread-0] 00:02:31.000000000 [message] [relaymiddle60~80.223.169.64] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,0,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
00:00:20.152356 [thread-0] 00:02:31.000000000 [message] [server25~74.125.22.191] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,0,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
00:00:20.152480 [thread-0] 00:02:31.000000000 [message] [server3~173.252.120.6] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,0,0,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0
00:00:20.152616 [thread-0] 00:02:31.000000000 [message] [4uthority1~11.0.1.116] [shd-tracker.c:440] [_tracker_logNode] [shadow-heartbeat] [node] 1,7957,6211,0.000000,0,0.000000;0,0,0,0,0,0,0,0,0,0,0,0;0,0,0,0,0,0,0,0,0,0,0,0;88,7957,73,4818,0,0,15,990,2149,0,0,0;77,6211,62,4092,0,0,15,990,1129,0,0,0
00:00:37.166044 [thread-0] n/a [message] [n/a] [shd-slave.c:390] [_slave_heartbeat] process resource usage at simtime 151587873627 reported by getrusage(): ru_maxrss=1.176247 GiB, ru_utime=0.183333 minutes, ru_stime=0.133333 minutes, ru_nvcsw=158292, ru_nivcsw=55174
```

Then I use gdb to attach to the shadow process. I inputted `continue` twice until the segfault occurs.
Here's the result of `continue` .
```
Attaching to process 2870
[New LWP 2872]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Thread 1 "shadow" received signal SIGTSTP, Stopped (user).
0x00007f137d2bc269 in raise (sig=20)
    at ../sysdeps/unix/sysv/linux/pt-raise.c:35
35	../sysdeps/unix/sysv/linux/pt-raise.c: No such file or directory.
(gdb) continue
Continuing.

Thread 2 "logger-helper" received signal SIGTSTP, Stopped (user).
[Switching to Thread 0x7f1377066700 (LWP 2872)]
clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:81
81	../sysdeps/unix/sysv/linux/x86_64/clone.S: No such file or directory.
(gdb) continue
Continuing.
[New Thread 0x7f13765fb700 (LWP 2894)]
[New Thread 0x7f1375dcf700 (LWP 2895)]
[New Thread 0x7f13755a3700 (LWP 2896)]
```

But something weird happened. I run gdb many times, the content behind the Segmentation fault are different. There are three results in total. I will post them all.
**1.This is the most frequent occurrence.**
```
Thread 5 "worker-2" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f3ac6fa7700 (LWP 2276)]
vdl_context_remove_file (context=0x0, file=file@entry=0x7f37edfcc150)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-context.c:274
274	  vdl_list_remove (context->loaded, file);
(gdb) backtrace
#0  vdl_context_remove_file (context=0x0, file=file@entry=0x7f37edfcc150)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-context.c:274
#1  0x00007f3ad3814552 in file_delete (mapping=true, file=0x7f37edfcc150)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-unmap.c:16
#2  vdl_unmap (files=files@entry=0x7f37ec050e50, mapping=mapping@entry=true)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-unmap.c:81
#3  0x00007f3ad380b3d3 in dlopen_with_context (context=0x7f382c779c50, 
    filename=<optimized out>, flags=<optimized out>)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-dl.c:345
#4  0x00007f382ae8f476 in ?? ()
#5  0x0000000000da4900 in ?? ()
#6  0x00007f3abedb0050 in ?? ()
#7  0x0000000000000001 in ?? ()
#8  0x00007f382ae913d3 in ?? ()
#9  0x1817171616151412 in ?? ()
#10 0x00007f382c431a19 in ?? ()
#11 0x00007f3abedb0050 in ?? ()
#12 0x0000000000000001 in ?? ()
#13 0x0000000000000000 in ?? ()
```
**2.It says `vdl_list_begin` and unexpectedly there have been two different results.**
2.1 The first result:
```
Thread 5 "worker-2" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f7c59dda700 (LWP 2515)]
vdl_list_begin (list=0x0)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-list.c:68
68	  read_lock (list->lock);
(gdb) backtrace
#0  vdl_list_begin (list=0x0)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-list.c:68
#1  0x00007f7c62109b03 in vdl_gc_white_list_new (
    list=list@entry=0x7f79800e2f50)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-gc.c:51
#2  0x00007f7c62109ce6 in vdl_gc_run ()
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-gc.c:124
#3  0x00007f7c621083b8 in dlopen_with_context (context=0x7f79899275b0, 
    filename=0x7f79822a426b "liblzma.so.2", flags=<optimized out>)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-dl.c:341
```

2.2 The second result:
```
Thread 5 "worker-2" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f4077fff700 (LWP 2307)]
vdl_list_begin (list=0x0)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-list.c:68
68	  read_lock (list->lock);
(gdb) backtrace
#0  vdl_list_begin (list=0x0)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-list.c:68
#1  0x00007f40845ac341 in vdl_sort_deps_breadth_first (
    file=file@entry=0x7f3e3302d590)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-sort.c:67
#2  0x00007f40845a4ca0 in vdl_dlvsym_with_flags (handle=<optimized out>, 
    symbol=0x47d6eb "__errno_location", version=0x0, flags=0, 
    caller=<optimized out>)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-dl.c:652
#3  0x0000000000435933 in _process_updateErrnoLocation (proc=0x1a1ea10)
    at /home/mason/Tor/shadow/src/main/host/shd-process.c:301
#4  0x0000000000435aa7 in _process_setErrno (proc=0x1a1ea10, errnoValue=17)
    at /home/mason/Tor/shadow/src/main/host/shd-process.c:323
#5  0x000000000043f26b in process_emu_open (proc=0x1a1ea10, 
    pathname=0x7f4066b206f0 "/etc/localtime", flags=0, mode=0)
    at /home/mason/Tor/shadow/src/main/host/shd-process.c:3388
```

**3.It's just appear once.**
```
Thread 4 "worker-1" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f6db9914700 (LWP 3019)]
vdl_lookup_file_begin (file=file@entry=0x7f6b19343fb0, 
    name=0x7f6b18aa5391 "memset", elf_hash=121391556, gnu_hash=226670640)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-lookup.c:99
99	      ElfW (Addr) bitmask = bitmask1 | bitmask2;
(gdb) backtrace
#0  vdl_lookup_file_begin (file=file@entry=0x7f6b19343fb0, 
    name=0x7f6b18aa5391 "memset", elf_hash=121391556, gnu_hash=226670640)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-lookup.c:99
#1  0x00007f6dc1419c65 in vdl_lookup_in_file (data=<optimized out>, 
    aux=0x7f6da58e7b60)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-lookup.c:452
#2  0x00007f6dc1419582 in vdl_list_search_on (list=list@entry=0x7f6b1a3980d0, 
    aux=aux@entry=0x7f6da58e7b60, 
    iterator=iterator@entry=0x7f6dc1419c20 <vdl_lookup_in_file>)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-list.c:532
#3  0x00007f6dc141a40e in vdl_lookup_with_scope_internal (
    scope=0x7f6b1a3980d0, args=0x7f6da58e7b60)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-lookup.c:533
#4  vdl_lookup (file=file@entry=0x7f6b19341c10, 
    name=name@entry=0x7f6b18aa5391 "memset", 
    ver_name=0x7f6b18aa57c3 "GLIBC_2.2.5", 
    ver_filename=0x7f6b18aa57a5 "libc.so.6", flags=<optimized out>)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-lookup.c:583
#5  0x00007f6dc141d855 in do_process_reloc (file=file@entry=0x7f6b19341c10, 
    reloc_type=7, reloc_addr=0x7f6b18cc5028, reloc_addend=0, reloc_sym=5)
    at /home/mason/Tor/shadow/src/external/elf-loader/vdl-reloc.c:111
#6  0x00007f6dc141dcb5 in process_rela (rela=<optimized out>, 
    file=0x7f6b19341c10)

```

Am I post the result clearly? Please let me know if there is anything I need to do.
Thanks.

**Comment by jtracey** [comment]
This looks to be the same bug as https://github.com/shadow/shadow/issues/423. I'm still not sure why this is happening in multiple instances on different distros, but none of my development machines or testing VMs. I'll do my best to write a fix tomorrow, and hopefully a test as well.

**Comment by jtracey** [comment]
this can be closed as a dup of shadow/shadow#423

**Comment by robgjansen** [comment]
Thanks @jtracey. Closing as dup of shadow/shadow#423.
